src/v/coproc/script_context_backend.cc

```diff
@@ -52,21 +48,37 @@ class set_term_id_to_zero {
     model::record_batch_reader::data_t _batches;
 };
 
+/// Necessary because the javascript project has potential areas where
+/// inconsistent record batches may be built and sent back to rp
+class verify_record_batch_consistency {
+public:
+    ss::future<ss::stop_iteration> operator()(const model::record_batch& b) {
+        _is_consistent &= (b.header().record_count == b.record_count());
+        co_return _is_consistent ? ss::stop_iteration::no
+                                 : ss::stop_iteration::yes;
+    }
+
+    bool end_of_stream() const { return _is_consistent; }
+
+private:
+    bool _is_consistent{true};
+};
+
 static ss::future<> do_write_materialized_partition(
   storage::log log, model::record_batch_reader reader) {
     /// Re-write all batch term_ids to 1, otherwise they will carry the
     /// term ids of records coming from parent batches
-    auto [success, batch_w_correct_terms]
+    auto [crc_success, header_complete, batch_w_correct_terms]
       = co_await std::move(reader).for_each_ref(
         reference_window_consumer(
-          model::record_batch_crc_checker(), set_term_id_to_zero()),
+          model::record_batch_crc_checker(),
+          verify_record_batch_consistency(),
+          set_term_id_to_zero()),
         model::no_timeout);
-    if (!success) {
-        /// In the case crc checks failed, do NOT write records to storage
-        throw crc_failed_exception(fmt::format(
-          "Batch failed crc checks, check wasm engine impl: {}",
-          log.config().ntp()));
-    }
+
+    vassert(crc_success, "Wasm engine failed to compile correct crc check");
+    vassert(header_complete, "Wasm engine returned malformatted batch/header");
```

Comment:
if I understand correctly this will crash-stop redpanda under scenarios like data corruption etc.. off the connection with the wasm node engine? if so we should probably consider containing the issue.

1. we observed at least one case where we believed a network security scanner touched the wasm rpc port and triggered a crc error. that seems like it would still be susceptible here, right?
2. if the error really is in the wasm engine, what good is it to crash redpanda process?

Reply:
1. No that didn't trigger a crc error, it just crashed the wasm engine, redpanda was fine, there was no communication between the engine and redpanda in that scenario.
2. Because if this occurs theres no good alternative. Retrys do not mean that eventually this problem will go away and resolve itself. If progress cannot be made then this will infinite loop forever if retry is the chosen alternative.

Reply:
> No that didn't trigger a crc error, it just crashed the wasm engine, redpanda was fine, there was no communication between the engine and redpanda in that scenario.

Right, but the generalization of that scenario is that if something poked the redpanda port the entire process would crash, right?

> Because if this occurs theres no good alternative. Retrys do not mean that eventually this problem will go away and resolve itself. If progress cannot be made then this will infinite loop forever if retry is the chosen alternative.

I'm not sure I follow. I believe you when you say there isn't a good alternative for dealing with this issue w.r.t. to making progress on the materialized topic, but not crash-stopping redpanda is a good alternative overall when the crash condition starts to depend on network connections and an external process like nodejs.

1. If redpanda is crash-stopped, will restarting resolve the issue?
2. If redpanda restarts, will it crash-loop when it attempts repeat processing?
3. Can coprocessor halt and report health errors until the issue resolved?

Overall I'm not opposed to crash stopping redpanda, but we need to be very clear about the full range of scenarios that might lead to the crash stop and the implications.

Reply:
2a. maybe just shutdown copro completely as a good alternative?

Reply:
Containing the fall out by shutting down copro and reporting health issues makes sense. But crash-stopping may also be reasonable. I think what is missing here is the analysis of the situation: what are the issues that lead to the crash, and what does the crash buy us? We need to think about it carefully because crash-stopping nodes with 100+ GB of RAM means a really long warm up period for the cache.

Reply:
redpanda/v/coproc is a client to a server which is the wasm engine, so one couldn't just make random requests to redpanda intending to hit copro stuff.

I agree, sounds like crashing might be a bit too extreme. I think we can log the event and deregister the coprocessor. That might be a good middle ground.

Reply:
when the health monitoring service is in place you could also probably publish information about a coprocessor failing into that central location for reporting. @mmaslankaprv?